google/resource_cloudfunctions_function.go

```diff
@@ -0,0 +1,462 @@
+package google
+
+import (
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"google.golang.org/api/cloudfunctions/v1"
+
+	"fmt"
+	"log"
+	"strings"
+	"time"
+)
+
+const DEFAULT_FUNCTION_TIMEOUT_IN_SEC = 60
+
+//Min is 1 second, max is 9 minutes 540 sec
+const FUNCTION_TIMEOUT_MAX = 540
+const FUNCTION_TIMEOUT_MIN = 1
+
+//Allowed values are: 128MB, 256MB, 512MB, 1024MB, and 2048MB. By default, a new function is limited to 256MB of memory.
+const FUNCTION_DEFAULT_MEMORY = 256
+
+var FUNCTION_ALLOWED_MEMORY = map[int]bool{
+	128:  true,
+	256:  true,
+	512:  true,
+	1024: true,
+	2048: true,
+}
+
+//For now CloudFunctions are allowed only in us-central1
+//Please see https://cloud.google.com/about/locations/
+var FUNCTION_ALLOWED_REGION = map[string]bool{
+	"us-central1": true,
+}
+
+func resourceCloudFunctionsFunction() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudFunctionsCreate,
+		Read:   resourceCloudFunctionsRead,
+		Update: resourceCloudFunctionsUpdate,
+		Delete: resourceCloudFunctionsDestroy,
+		Importer: &schema.ResourceImporter{
+			State: func(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+				d.Set("name", d.Id())
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(5 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Delete: schema.DefaultTimeout(5 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+
+			"entry_point": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+
+			"memory": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeInt},
+				Default:  FUNCTION_DEFAULT_MEMORY,
+			},
+
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+
+			"timeout": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				Default:      DEFAULT_FUNCTION_TIMEOUT_IN_SEC,
+				ValidateFunc: validation.IntBetween(FUNCTION_TIMEOUT_MIN, FUNCTION_TIMEOUT_MAX),
+			},
+
+			"labels": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+
+			"storage_bucket": {
+				Type:     schema.TypeString,
+				Required: true,
```

Comment:
We typically sort our fields in the schema in the order Required -> Optional -> Computed- it makes it a little easier to navigate.

Reply:
Done that :)